### DISCUSSION TITLE
How can i get the actual dpi of a picture in pdf

### DISCUSSION THREAD

**Discussion by R4Y-R4Y** [pymupdf/PyMuPDF#1801] [Looking for help] [answered]
Hello
so i am using PyMuPDF to extract pictures for ocr scanning
i want to check the dpi of pictures so i wont scan pictures that are basically unusable for tesseract
from what i saw in the docs there's `pic.xres` or `pic.yres` but it's always by default 96
I've tested it with two pictures in documents but its always the same
can someone please help me :)

**Comment by JorjMcKie** [comment, marked as answer]
I hope you are extracting the "pictures" via `doc.extract_image(xref)`. Given this, if you make a `Pixmap` of the image or, equivalently, a Pillow `Image`, then you will get access to the DPI value stored in it. This is all that is available. Depending on the packages, a few default assumptions are made if no, or an invalid DPI is stored.

**Reply by R4Y-R4Y** [reply]
oh 
i am extracting the images of each page  with `doc.get_pages_images(page)` can you tell me what's the difference between them?

**Reply by R4Y-R4Y** [reply]
just to check here's the code that goes to every page in the pdf ans saves it into a variable and saves the `pic.xres` and `pic.yres`

for j in range(pdf.page_count):
    img_pdf = pdf.get_page_images(j)
    for image_page in img_pdf:
        xref = image_page[0]
        res = fitz.Pixmap(pdf, xref)
        dpix , dpiy = res.xres , res.yres
        print(dpix, ' ', dpiy)
        x = pix2np(res)

**Reply by JorjMcKie** [reply]
Ah, that's good enough too. But try with the other method to be sure.
Do you have reason to believe the result cannot be true? You may want to crosscheck by extracting the images to separate files and skim through them with some tool.

**Reply by JorjMcKie** [reply]
and also make a Pillow image from it and crosscheck Pillow's result.

**Reply by JorjMcKie** [reply]
like this
```python
from PIL import Image
import io
img=doc.extract_image(xref)
fp = io.BytesIO(img["image"])
pilimg = Image.open(fp)
print(pilimg.info["dpi"])
```

**Reply by R4Y-R4Y** [reply]
and can i convert this to numpy array since opencv uses only numpy arrays when it comes to editing pictures?

**Reply by JorjMcKie** [reply]
The `Pixmap` object contains property `.samples`  which represents a rectangular area of bytes of this shape:
width * height * 3, where 3 equals the RGB number of bytes per pixel.
So you easily can create a numpy ndarray from this.

**Reply by R4Y-R4Y** [reply]
okay so i have tested this with this do [here](https://github.com/pymupdf/PyMuPDF/files/9065627/dtest.pdf)
here's the current code :
```python
def scanned_dc(pdf):
    for j in range(pdf.page_count):
            img_pdf = pdf.get_page_images(j)
            for image_page in img_pdf:
                xref = image_page[0]
                res = pdf.extract_image(xref)
                fp = io.BytesIO(res["image"])
                x = Image.open(fp)
                print(x.info["dpi"])
```
and i got this error:
```python
File "********************************************************************", line 34, in scanned_dc
    print(x.info["dpi"])
KeyError: 'dpi'
```
idk what to do :)

**Reply by R4Y-R4Y** [reply]
also it worked with another doc and it said it had 72 dpi instead of the default 96 but couldn't convert it to numpy with this function:
```python
def pix2np(pix):
    """
    Converts a pixmap buffer into a numpy array
    """
    # pix.samples = sequence of bytes of the image pixels like RGBA
    #pix.h = height in pixels
    #pix.w = width in pixels
    # pix.n = number of components per pixel (depends on the colorspace and alpha)
    im = np.frombuffer(pix.samples, dtype=np.uint8).reshape(
        pix.h, pix.w, pix.n)
    try:
        im = np.ascontiguousarray(im[..., [2, 1, 0]])  # RGB To BGR
    except IndexError:
        # Convert Gray to RGB
        im = cv2.cvtColor(im, cv2.COLOR_GRAY2RGB)
        im = np.ascontiguousarray(im[..., [2, 1, 0]])  # RGB To BGR
    return im
```
it gives me this error:
```python
  File "****************************************************************", line 16, in pix2np
    im = np.frombuffer(pix.samples, dtype=np.uint8).reshape(
AttributeError: 'dict' object has no attribute 'samples'
```
the stars is just for hiding the files

**Reply by JorjMcKie** [reply]
> 'dict' object has no attribute 'samples'

**_Obviously_** pix was not a `Pixmap`!

**Reply by R4Y-R4Y** [reply]
yeah i found out how to change it by simply changing its type to np.array `x = np.array(x)`
thanks for your help you're really a savior and helped me a lot :)

**Reply by JorjMcKie** [reply]
pleased to hear that!

**Comment by JorjMcKie** [comment]
> idk what to do :)

neither do I, without any way to reproduce the situation.

**Reply by JorjMcKie** [reply]
Ah got the file.
Yes that image has no embedded dpi info. That dict is empty. If this occurs (use try/except to catch), assume 96 like (Py-) MuPDF does.

**Reply by R4Y-R4Y** [reply]
oh that's interesting
so its value is `None` or sth like that?

**Reply by JorjMcKie** [reply]
No the PIL image has `image.info = {}`.

**Reply by R4Y-R4Y** [reply]
> No the PIL image has `image.info = {}`.

so basically an empty dictionary
cool
thanks for the support :)

**Comment by Abhishek-choukuse** [comment]
hey, I had the same doubt here. how did you exactly find the dpi of an image based pdf then?
It would be really helpful if you could explain what you did exactly

**Reply by robinwatts** [reply]
PDF files contain images. As far as PDF is concerned these images are just a 2d array of pixels - so we know the width/height, and the values at each point. That's all.

Any information they may contain as to their dpi is completely irrelevant - it's not used.

The PDF graphic operators will set up a transform to display the images; scaling every image from a conceptual unit square onto the page.

So the 'dpi' of an image is not to do just with the image itself, but rather is to do with the way in which the image is used. (The same image might get used multiple times within the document with different scales; the dpi will thus differ per instance).

So, to find the 'dpi' of a scanned document, you need to look both at the dimensions of the image, and how the image is used on the page.

**Reply by JorjMcKie** [reply]
You must make up your mind what you want to see under the label "dpi": **_either_** the values baked into the image binary file (which is what I did above), **_or_** the ratio image-width (height) divided by the bbox-width (height) on the page, multiplied by 72.

**Reply by JorjMcKie** [reply]
Ah, @robinwatts answer is more enlightening. Just worked in parallel.

**Reply by Abhishek-choukuse** [reply]
@JorjMcKie  
The values baked into the image binary file in my opinion contains resolution at which the image is being displayed/stored in the PDF. not the resolution at which the document was originally scanned.
For example, it shows 300 DPI baked in the image binary file, but if I use online available tools to calculate the dpi of the doc, it shows 116. and it indeed is less than 150 because I tried OCR on this doc with DPI 150 and the file size actually increased, indicating DPI should be less than 150.

**Reply by JorjMcKie** [reply]
> The values baked into the image binary file in my opinion contains resolution at which the image is being displayed/stored in the PDF. not the resolution at which the document was originally scanned.

No! See @robinwatts comments. When an image is embedded in a PDF, its baked-in DPI values are not touched.

Most tools showing DPI values for images displayed in a PDF make a calculation like my second one above:
* x-resolution = image-width / bbox-width * 72
* y-resolution = image-height / bbox-height * 72

This means that these values can vary per image in the same PDF. For instance. when the image is displayed multiple times inside bboxes of different sizes.

**Reply by Abhishek-choukuse** [reply]
does this mean x-resolution and y-resolution can be different? if so, what does it mean exactly? because all the tools that I have used till now to find the DPI of a doc gives only 1 DPI. so is that average of x-res and y-res?

**Reply by JorjMcKie** [reply]
Of course they can be different: just insert an image and scale it differently in x and y direction.
I don't know what those tools do. Different resolution values per dimension are not used often.
Original:
![image](https://github.com/user-attachments/assets/05232586-77aa-41c1-ba53-da7c3267eee2)
Inserted:
![image](https://github.com/user-attachments/assets/9a7504ee-7f94-442f-b52d-66227660aa3f)

**Reply by JorjMcKie** [reply]
Example tool: https://apitemplate.io/pdf-tools/free-pdf-dpi-analyzer/
Results for this PDF:
![image](https://github.com/user-attachments/assets/a3209c78-264a-4882-822e-a24cc8b81ac2)
Confirm the values yourself:
* x: 439 / 200 * 72 = 158.04
* y: 501 / 100 * 72 = 360.72
* the baked-in resolutions are 
![image](https://github.com/user-attachments/assets/10c43c68-51a4-46ce-be3f-bad90364b78c)

[x.pdf](https://github.com/user-attachments/files/20901629/x.pdf)
